More Local Dev Server Support (#3252)
googleforgames/agones: site/content/en/docs/Guides/Client SDKs/local.md

will not try to connect to anything, and will just send log messages to stdout and persist local state in memory so
that you can see exactly what the SDK in your game server is doing, and can
confirm everything works.

To do this you will need to download {{% ghrelease file_prefix="agonessdk-server" %}}, and unzip it.
For local development with integration into your cluster, the SDK Server can also be run in "out of cluster mode", discussed more [below](#running-locally-using-out-of-cluster-mode).

## Running the SDK Server

To run the SDK Server, you will need a copy of the binary.
This can either be done by downloading a prebuilt binary or running from source code.

### Getting the prebuilt binary

To run the prebuilt binary, for the latest release, you will need to download {{% ghrelease file_prefix="agonessdk-server" %}}, and unzip it.
You will find the executables for the SDK server, for each type of operating system.
`sdk-server.darwin.amd64` and `sdk-server.darwin.arm64` are for <u>MacOS</u>, `sdk-server.linux.amd64` and `sdk-server.linux.arm64` are for <u>Linux</u>, and `sdk-server.windows.amd64.exe` is for <u>Windows</u>.

> macOS

Reviewer: I'd prefer to keep this section in. Not everyone knows what OS matches to what code names, etc. Better to be explicit.

Author: This section was still here. It's right above this. I just changed this to be a formatted sentence, rather than half-bullets. I went ahead and changed this mostly back (although, now its a single set of bullets).

> ### Running from source code

Author: Moved to last section (before `Next Steps`).

> Instead of downloading and running executable binaries from the internet, you can simply run from source code.

Reviewer: 🤔 not sure we want to be advocating people compile from source. I mean you _can_, not sure if you _should_ ? It feels like it complicates things quite a bit, and there is no real reason to  - we provide binaries (and docker images too)?

Unless there is a strong reason to that I'm missing? (I can't think of another project that does this).

Removing the documentation on compilation would also simplify everything considerably.

Author: Compiling from source is a very common occurrence for handling dependencies. For example, Bazel makes use of pulling source tar balls from dependency libs. Gentoo replaces its entire package manager with downloading and compiling source code. Many code bases also offer this as an alternative to downloading prebuilt binaries.

Downloading unknown binaries, especially with no SHA256, is decently frowned upon in many places. I definitely was apprehensive about "here, run this exe". FWIW, interpreted code (straight from source) or docker images don't really have this issue because you see exactly what is being run.

Honestly, building from source is really helpful -- especially with interpreted languages (or languages that "feel interpreted", https://go.dev/doc/). It's really easy to test, debug, iterate, etc. I understand that for most binaries in Agones, this doesn't really make sense -- one can't run from source for a controller in a cluster. But for the SDK Server, when run locally, this actually makes a lot of sense.

Reviewer: This says two things to me:

1. We should file a ticket to have SHA's with our binaries - that's a good point to make.
2. Maybe we need to improve our documentation on using the container based approach?

This particular documentation's focus is on local testing - not on compiling from source (maybe that should be elsewhere - would like more eyes on that before going down that path), so conflating the two makes things unnecessarily complicated.

I'd say remove it. If you know how to compile source, you know how to go down that path -- if not, it's teaching git, and go and everything else, all to do local testing, and that's a lot.

I don't disagree on the incentives to compile from source - but I think removing the documentation for it in this section is worth the better onboarding and developer documentation experience.

Calling in some extra eyes to see if I'm off base here: @zmerlynn @igooch @gongmax - WDYT?

Author: > We should file a ticket to have SHA's with our binaries - that's a good point to make.

That definitely sounds like a good idea.

> Maybe we need to improve our documentation on using the container based approach?

I think container based approaches for dev workflows have pros and cons. The major pros are that all you need is docker installed; after that, everything is just a few commands to get running. The downside is that its more levels of stuff to run/debug. If one is wanting to delve into Agones (like me 😄) or just wanting to have a simpler dev environment, I don't think forcing people into larger constructs (like containerization) is the best.

> This particular documentation's focus is on local testing - not on compiling from source

I don't particularly see the diff here. But, as I said in one of my other comments, I'm fine with changing bash samples over to use the downloaded binaries (for consistency).

> I'd say remove it. If you know how to compile source, you know how to go down that path

I still think that it's nice to at least nudge people in the right direction. When I read through this page, I just wanted to run from source. I had to then dig around the source code to find the file to run. If the documentation just said "you want `cmd/sdk-server/main.go`", that would have been amazingly helpful on its own.

> if not, it's teaching git, and go and everything else, all to do local testing, and that's a lot.

I don't think its the documentation's job to ensure the reader understands prerequisite info. Just like other pages aren't responsible for teaching the reader what REST is or how `curl` works. 🤷 Running a game server service on k8s is not trivial. If the reader cant run from source code, doesn't know what git is (on GitHub), etc, they're in a bit of a rut.

Reviewer: One extra thought I had as a probable good middle ground (and in combination with the other comment I added to `local-game-server.md`):

1. Move most of the documentation for how to run "out of cluster" into `local-game-server.md`, but do the docs with running as a binary (assuming you agree they go better together as a development flow), as that is the general explanation of "how do I do an out of cluster setup"
2. On this page, we already have a "Running Local Mode in a Container", so why don't we add a section right below that of "Running Local Mode from source code" - where we _can_ explain how to run from source (and if need be explain how to git clone the right release branch etc).

That way we aren't conflating running from source with how to do other things, but if you want the source path, then you can definitely go down that path, and the docs for how to use the local sdk can just use the binary for simplicity.

WDYT?

Reviewer 2: I'm generally in agreement with @markmandel that running from source is not the direction I would prefer to support on the public docs. The main reason is just that: support. Documenting a path comes with a certain "warranty", implied or not, and I can easily see someone filing a bug because e.g. `go run` didn't work for them due to differing Go versions. I I don't personally want to be in the business of supporting every user's Go environ and understanding how it differs from our build image.

I'm not against bumping this out to a section with a disclaimer, though.

Author: > Move most of the documentation for how to run "out of cluster" into `local-game-server.md`

Maybe this is a bit of a duplicated topic, I also wrote in https://github.com/googleforgames/agones/pull/3252#discussion_r1264965321 that I think it _could_ go in either place. But here made more sense as this is where the user is changing what is running (different cli args).

> On this page, we already have a "Running Local Mode in a Container", so why don't we add a section right below that of "Running Local Mode from source code" - where we can explain how to run from source (and if need be explain how to git clone the right release branch etc).

That sounds good to me. I can re-arrange the sections to make it more linear/separated out.

> I don't personally want to be in the business of supporting every user's Go environ and understanding how it differs from our build image.
> I'm not against bumping this out to a section with a disclaimer, though.

I like disclaimers. 😄 I think this is a very understandable stance to take, I'd be happy to add in a "this is here to document that it can be done; you're on your own for ensuring your golang environment is setup properly".

> Instead of downloading and running executable binaries from the internet, you can simply run from source code.
> First clone the [Agones GitHub repo](https://github.com/googleforgames/agones).
> You can switch to a specific release's branch/tag or run from main.
> Running from source code will require having golang installed, which can be done by following the instructions [here](https://go.dev/doc/install).

Reviewer: ```suggestion
If you wish to run from source rather than pre-built binaries,  you will need [Go installed](https://go.dev/doc/install).

**Disclaimer:** Agones is run and tested with the version of Go specified by the `GO_VERSION` variable in the project's [build Dockerfile](https://github.com/googleforgames/agones/blob/main/build/build-image/Dockerfile). Other versions are not supported, but may still work.

You will need to clone the [Agones GitHub repo](https://github.com/googleforgames/agones) and switch to your installed specific release's branch/tag. For example:

```bash
git clone https://github.com/googleforgames/agones.git
cd agones
git checkout release-{{< release-version >}}
```

```

Adding an example and simplifying the language (although I think my suggestion formatting in here got a bit weird).

Author: I've moved "running from source" down to the last section (just above `Next Step`). Reworded to be a mix of new and old. Feel free to suggest further edits.

> With golang installed and the Agones repository cloned, the SDK Server can easily be run with the following command (from the agones clone directory):
> ```bash
> go run cmd/sdk-server/main.go
> ```
> Note: This command does not contain any of the necessary command line flags used in the following sections.
> It simply serves as an example of how to run from source code instead of running the prebuilt binary.
> Passing commandline flags (e.g. `--local`) will work in the same way.

Reviewer: ```suggestion
With Go installed and the Agones repository cloned, the SDK Server can be run with the following command (from the Agones clone directory):
```bash
go run cmd/sdk-server/main.go --local
/```

Commandline flags (e.g. `--local`) are exactly the same as command line flags when utilising a pre-built binary.
```

Tech writer nit: Some things aren't "easily" or "simple" for some people, so better to remove the adjective.

Author: Cleaned up.

```bash
./sdk-server.linux.amd64 --local
```

> or
> ```bash
> go run cmd/sdk-server/main.go --local
> ```

Reviewer: ```suggestion
```

Since we have the section above now.

Author: Cleaned up any `go run` outside of the "run from source" section.

chmod o+r gameserver.yaml
docker run --network=host --rm -v $(pwd)/gameserver.yaml:/tmp/gameserver.yaml us-docker.pkg.dev/agones-images/release/agones-sdk:{{<release-version>}} --local -f /tmp/gameserver.yaml
```

> 

Reviewer: The same as our other docs: https://github.com/googleforgames/agones/blob/3d57eaa4fe18590634c17137e369302c981b91b9/site/content/en/docs/Integration%20Patterns/allocation-from-fleet.md?plain=1#L28

```suggestion

## Next Steps:
* Learn how to connect your local development game server binary into a running Agones Kubernetes cluster for even more live [development options](({{< ref "/docs/Guides/local-game-server.md" >}})).

```

(This would be at the bottom of the page -- see note on moving below to `local-game-server.md`).

Author: Added/updated `Next Steps` on `site/content/en/docs/Guides/Client SDKs/local.md` and `site/content/en/docs/Guides/local-game-server.md` to link to the new `site/content/en/docs/Advanced/out-of-cluster-dev-server.md` page.

> ## Running locally using "out of cluster" mode

Reviewer: Just a heads up, our docs publish on merge, so if you have new features that aren't in the current release, please wrap them in a `feature` shortcode like so: https://agones.dev/site/docs/contribute/

Author: Added `publishDate` to new page. Wrapped links with `{{% feature publishVersion="1.34.0" %}}`.

An alternative to running completely isolated from a cluster is to run in "out of cluster" mode.
This allows you to run locally but interact with the controllers within a cluster.

> This workflow works well when running a [Local Game Server](https://agones.dev/site/docs/guides/local-game-server/) in your cluster and want to run the server locally.

Reviewer: Rather than use links to the website, please either the `ref` or `relref` shortcodes.

Author: Good call. Sorry. Will update.

This means being able to allocate your game in its normal flow (much more prod-like environment) and be able to debug (e.g. breakpoint) your server code.
This can also be done with [running Minikube locally](https://agones.dev/site/docs/installation/creating-cluster/minikube/), which is great for early prototyping.

The name "out of cluster" mode is to contrast [InClusterConfig](https://pkg.go.dev/k8s.io/client-go/tools/clientcmd#InClusterConfig) which is used in the internal golang kubeconfig API.

To run in "out of cluster" mode, instead of passing `--local` or `-f ./gameserver.yaml`, you'd use `--kubeconfig` to connect into your cluster.
However, there are a number of commands that are necessary/useful to run when running in "out of cluster" mode.
Here is a sample with each piece discussed after.
```bash

> GAMESERVER_NAME='my-local-server' \

Reviewer: I feel like to do this cleanly, we should update the sdkserver to also allow command line args for the GameServer name and Pod Namespace -- just to be consistent, and make life easier for end users. Then we don't have to mix args and env vars.

Also would prefer to see these docs using a binary, not running from source.

Author: > Then we don't have to mix args and env vars.

I agree that this would be nicer. I can add in the extra flags.

> Also would prefer to see these docs using a binary, not running from source.

As mentioned in the other comment, I think running from source is nicer than downloaded binaries. But sure, for the sake of documentation consistency, I can update this.

Author: Just to follow up here, this work has been done. `--gameserver-name` and `--pod-namespace` can be used as cli args, and `GAMESERVER_NAME` and `POD_NAMESPACE` still work as env vars (which are used in k8s). Also updated this to be consistent and use pre-built binary commands.
